libraries/lib-uuid/Uuid.cpp

```diff
@@ -0,0 +1,216 @@
+              std::string::const_iterator& inputIt,
+              const std::string::const_iterator& inputEnd)
+{
+   if (inputIt == inputEnd)
+      return false;
+
+   const char c1 = *inputIt++;
+
+   if (!std::isxdigit(c1))
+      return false;
+
+   if (inputIt == inputEnd)
+      return false;
+
+   const char c2 = *inputIt++;
+
+   if (!std::isxdigit(c2))
+      return false;
+
+   *outputIt = uint8_t((HexCharToNum(c1) << 4) | HexCharToNum(c2));
+
+   ++outputIt;
+
+   return true;
+}
+} // namespace
+
+Uuid::Uuid()
+    : Uuid(Bytes {})
+{
+}
+
+Uuid::Uuid(const Bytes& data) noexcept
+    : mData(data)
+{
+}
+
+Uuid Uuid::Generate()
+{
+#if defined(USE_UUID_CREATE)
+   UUID winUid;
+
+   if (RPC_S_OK != ::UuidCreate(&winUid))
+      return {};
+
+   Uuid uuid;
+
+   std::memcpy(uuid.mData.data(), &winUid, sizeof(winUid));
+
+   return uuid;
+#elif defined(USE_CFUUID)
+   CFUUIDRef newId = CFUUIDCreate(NULL);
+   CFUUIDBytes bytes = CFUUIDGetUUIDBytes(newId);
+   CFRelease(newId);
+
+   Uuid uuid;
+
+   std::memcpy(uuid.mData.data(), &bytes, sizeof(bytes));
+
+   return uuid;
+#elif defined(USE_LIBUUID)
+   uuid_t newId;
+
+   uuid_generate(newId);
+
+   Uuid uuid;
+
+   std::memcpy(uuid.mData.data(), newId, sizeof(newId));
+
+   return uuid;
+#else
+#   error "UUID generator is not defined"
+#endif
+}
+
+Uuid Uuid::FromString(const std::string& str)
+{
+   const size_t length = str.length();
+
+   if (length == 0)
+      return {};
+
+   const bool hasBraces = str[0] == '{';
+
+   if (hasBraces && (length != 38 || str.back() != '}'))
+      return {};
+   else if (!hasBraces && length != 36)
+      return {};
+
+   std::string::const_iterator currentSymbol =
+      hasBraces ? str.begin() + 1 : str.begin();
+   std::string::const_iterator inputEnd = str.end();
+
+   Uuid uuid;
+   Bytes::iterator currentByte = uuid.mData.begin();
+
+   for (int i = 0; i < 16; ++i)
+   {
+      if (!readByte(currentByte, currentSymbol, inputEnd))
+         return {};
+
+      if (currentSymbol != inputEnd && *currentSymbol == '-')
+         ++currentSymbol;
+   }
+
+   return uuid;
+}
+
+bool Uuid::isNil() const noexcept
+{
+   return std::all_of(mData.begin(), mData.end(),
+                      [](uint8_t c) { return c == 0; });
+}
+
+bool Uuid::operator==(const Uuid& rhs) const noexcept
+{
+   return mData == rhs.mData;
+}
+
+bool Uuid::operator!=(const Uuid& rhs) const noexcept
+{
+   return mData != rhs.mData;
+}
+
+bool Uuid::operator>(const Uuid& rhs) const noexcept
+{
+   return mData > rhs.mData;
+}
+
+bool Uuid::operator>=(const Uuid& rhs) const noexcept
+{
+   return mData >= rhs.mData;
+}
+
+bool Uuid::operator<(const Uuid& rhs) const noexcept
+{
+   return mData < rhs.mData;
+}
+
+bool Uuid::operator<=(const Uuid& rhs) const noexcept
+{
+   return mData <= rhs.mData;
+}
+
+Uuid::Bytes Uuid::toBytes() const
+{
+   return mData;
+}
+
+std::string Uuid::toString() const
+{
+   char buffer[36 + 1] = {};
+
+   snprintf(
```

Comment:
I might assert that return value is 36
Then define a constant so I don't repeat 36